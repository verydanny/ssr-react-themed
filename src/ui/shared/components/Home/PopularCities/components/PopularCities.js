import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import themed from 'react-themed'
import PopularLinks from '../../PopularLinks'

@themed(/^HomePopularCities/, { pure: true })
export default class PopularCities extends PureComponent {
  static propTypes = {
    content: PropTypes.shape({
      headline: PropTypes.string,
      subHeadline: PropTypes.string,
      apartmentsLinks: PropTypes.string,
      housesLinks: PropTypes.string,
    }),
    theme: PropTypes.object,
  }

  static defaultProps = {
    content: {},
    theme: {},
  }

  state = {
    parsedApartmentsLinks: null,
    parsedHousesLinks: null,
  }

  componentDidMount() {
    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState({
      parsedApartmentsLinks: this.parseLinks(this.apartmentsRef),
      parsedHousesLinks: this.parseLinks(this.housesRef),
    })
  }

  get apartmentsLinks() {
    const { content } = this.props

    return this.renderLinksSection('apartmentsRef', content.apartmentsLinks)
  }

  get housesLinks() {
    const { content } = this.props

    return this.renderLinksSection('housesRef', content.housesLinks)
  }

  parseLinks(ref) {
    if (!ref) return null

    const elements = ref.querySelectorAll('a')

    return Array.from(elements).map(anchor => ({
      url: anchor.href,
      displayName: anchor.textContent,
    }))
  }

  renderLinksSection(id, html) {
    /* eslint-disable react/no-danger */
    return (
      <div
        dangerouslySetInnerHTML={{ __html: html }}
        ref={ref => { this[id] = ref }}
      />
    )
  }

  renderLinks(heading, links) {
    const { theme } = this.props

    return (
      <div className={theme.HomePopularCities_Section}>
        <h4>{heading}</h4>
        {links}
      </div>
    )
  }

  render() {
    const { content, theme } = this.props
    const {
      headline,
      subHeadline,
      apartmentsLinks,
      housesLinks,
    } = content
    const { parsedApartmentsLinks, parsedHousesLinks } = this.state

    if (!headline || !apartmentsLinks || !housesLinks) return null

    if (parsedApartmentsLinks || parsedHousesLinks) {
      return (
        <PopularLinks
          headline={headline}
          subHeadline={subHeadline}
          apartmentsLinks={parsedApartmentsLinks}
          housesLinks={parsedHousesLinks}
          tagSection="popular_cities"
          testId="home-popular-cities"
        />
      )
    }

    // This needs to rendered as raw HTML, once server-side (for SEO) and once
    // client-side so we can parse the links and transform into our current design.
    return (
      <div
        className={theme.HomePopularCities}
        data-tid="home-popular-cities"
        data-tag_section="popular_cities"
      >
        <h2>{headline}</h2>
        {subHeadline && <h3>{subHeadline}</h3>}
        {this.renderLinks('Apartments', this.apartmentsLinks)}
        {this.renderLinks('Houses', this.housesLinks)}
      </div>
    )
  }
}
