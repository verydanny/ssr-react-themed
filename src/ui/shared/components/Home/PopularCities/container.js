import { connect } from 'react-redux'
import { getHomePopularCities } from 'app/selectors/meta'
import PopularCities from './components/PopularCities'

const mapStateToProps = state => ({
  content: getHomePopularCities(state),
})

export default connect(mapStateToProps, {})(PopularCities)
