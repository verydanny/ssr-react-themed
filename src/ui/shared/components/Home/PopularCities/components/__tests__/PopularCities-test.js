import React from 'react'
import { shallow, mount } from 'enzyme'
import PopularCities from '../PopularCities'

const apartmentsTestLinks = [
  { url: '/georgia/atlanta-apartments', text: 'Atlanta, GA' },
  { url: '/north-carolina/charlotte-apartments', text: 'Charlotte, NC' },
  { url: '/illinois/chicago-apartments', text: 'Chicago, IL' },
  { url: '/california/los-angeles-apartments', text: 'Los Angeles, CA' },
  { url: '/new-york/new-york-apartments', text: 'New York, NY' },
  { url: '/michigan/detroit-apartments', text: 'Detroit, MI' },
  { url: '/california/san-diego-apartments', text: 'San Diego, CA' },
  { url: '/texas/dallas-apartments', text: 'Dallas, TX' },
]

const housesTestLinks = [
  { url: '/georgia/atlanta-houses', text: 'Atlanta, GA' },
  { url: '/north-carolina/charlotte-houses', text: 'Charlotte, NC' },
  { url: '/illinois/chicago-houses', text: 'Chicago, IL' },
  { url: '/california/los-angeles-houses', text: 'Los Angeles, CA' },
  { url: '/new-york/new-york-houses', text: 'New York, NY' },
  { url: '/michigan/detroit-houses', text: 'Detroit, MI' },
  { url: '/california/san-diego-houses', text: 'San Diego, CA' },
  { url: '/texas/dallas-houses', text: 'Dallas, TX' },
]

const apartmentsLinks = `
  <ul>
    <li><a href="${apartmentsTestLinks[0].url}">${apartmentsTestLinks[0].text}</a></li>
    <li><a href="${apartmentsTestLinks[1].url}">${apartmentsTestLinks[1].text}</a></li>
    <li><a href="${apartmentsTestLinks[2].url}">${apartmentsTestLinks[2].text}</a></li>
    <li><a href="${apartmentsTestLinks[3].url}">${apartmentsTestLinks[3].text}</a></li>

    <li><a href="${apartmentsTestLinks[4].url}">${apartmentsTestLinks[4].text}</a></li>
    <li><a href="${apartmentsTestLinks[5].url}">${apartmentsTestLinks[5].text}</a></li>
    <li><a href="${apartmentsTestLinks[6].url}">${apartmentsTestLinks[6].text}</a></li>
    <li><a href="${apartmentsTestLinks[7].url}">${apartmentsTestLinks[7].text}</a></li>
  </ul>
`

const housesLinks = `
  <ul>
    <li><a href="${housesTestLinks[0].url}">${housesTestLinks[0].text}</a></li>
    <li><a href="${housesTestLinks[1].url}">${housesTestLinks[1].text}</a></li>
    <li><a href="${housesTestLinks[2].url}">${housesTestLinks[2].text}</a></li>
    <li><a href="${housesTestLinks[3].url}">${housesTestLinks[3].text}</a></li>

    <li><a href="${housesTestLinks[4].url}">${housesTestLinks[4].text}</a></li>
    <li><a href="${housesTestLinks[5].url}">${housesTestLinks[5].text}</a></li>
    <li><a href="${housesTestLinks[6].url}">${housesTestLinks[6].text}</a></li>
    <li><a href="${housesTestLinks[7].url}">${housesTestLinks[7].text}</a></li>
  </ul>
`

const content = {
  headline: 'Popular Cities Headline',
  subHeadline: 'Popular Cities SubHeadline',
  apartmentsLinks,
  housesLinks,
}

const props = {
  content,
  initialVisibleLinks: 2,
}

const context = {
  theme: {
    HomePopularCities_Section: 'section',
    HomePopularLinks_Section: 'section',
  },
}

describe('ui/shared/components/Home/PopularCities', () => {
  it('server renders apartments HTML', () => {
    const wrapper = shallow(<PopularCities {...props} />, { context })
    const apartments = wrapper.render().find('.section').eq(0)
    expect(apartments.html()).toMatchSnapshot()
  })

  it('server renders houses HTML', () => {
    const wrapper = shallow(<PopularCities {...props} />, { context })
    const houses = wrapper.render().find('.section').eq(1)
    expect(houses.html()).toMatchSnapshot()
  })

  it('parses correct apartments links', () => {
    const wrapper = mount(<PopularCities {...props} />, { context })
    const apartments = wrapper.find('.section').at(0)
    expect(apartments.html()).toMatchSnapshot()
  })

  it('parses correct houses links', () => {
    const wrapper = mount(<PopularCities {...props} />, { context })
    const houses = wrapper.find('.section').at(1)
    expect(houses.html()).toMatchSnapshot()
  })
})
