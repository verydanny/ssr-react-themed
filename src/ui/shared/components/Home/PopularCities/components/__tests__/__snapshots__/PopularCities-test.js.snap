// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/shared/components/Home/PopularCities parses correct apartments links 1`] = `

<div class="section">
  <h4>
    Apartments
  </h4>
  <div class>
    <div data-tid="home-popular-cities-apartments">
      <div>
        <a href="/georgia/atlanta-apartments"
           data-tag_item="apartments_link"
        >
          Atlanta, GA
        </a>
        <a href="/north-carolina/charlotte-apartments"
           data-tag_item="apartments_link"
        >
          Charlotte, NC
        </a>
        <a href="/illinois/chicago-apartments"
           data-tag_item="apartments_link"
        >
          Chicago, IL
        </a>
        <a href="/california/los-angeles-apartments"
           data-tag_item="apartments_link"
        >
          Los Angeles, CA
        </a>
      </div>
      <div>
        <a href="/new-york/new-york-apartments"
           data-tag_item="apartments_link"
        >
          New York, NY
        </a>
        <a href="/michigan/detroit-apartments"
           data-tag_item="apartments_link"
        >
          Detroit, MI
        </a>
        <a href="/california/san-diego-apartments"
           data-tag_item="apartments_link"
        >
          San Diego, CA
        </a>
        <a href="/texas/dallas-apartments"
           data-tag_item="apartments_link"
        >
          Dallas, TX
        </a>
      </div>
    </div>
  </div>
</div>

`;

exports[`ui/shared/components/Home/PopularCities parses correct houses links 1`] = `

<div class="section">
  <h4>
    Houses
  </h4>
  <div class>
    <div data-tid="home-popular-cities-houses">
      <div>
        <a href="/georgia/atlanta-houses"
           data-tag_item="houses_link"
        >
          Atlanta, GA
        </a>
        <a href="/north-carolina/charlotte-houses"
           data-tag_item="houses_link"
        >
          Charlotte, NC
        </a>
        <a href="/illinois/chicago-houses"
           data-tag_item="houses_link"
        >
          Chicago, IL
        </a>
        <a href="/california/los-angeles-houses"
           data-tag_item="houses_link"
        >
          Los Angeles, CA
        </a>
      </div>
      <div>
        <a href="/new-york/new-york-houses"
           data-tag_item="houses_link"
        >
          New York, NY
        </a>
        <a href="/michigan/detroit-houses"
           data-tag_item="houses_link"
        >
          Detroit, MI
        </a>
        <a href="/california/san-diego-houses"
           data-tag_item="houses_link"
        >
          San Diego, CA
        </a>
        <a href="/texas/dallas-houses"
           data-tag_item="houses_link"
        >
          Dallas, TX
        </a>
      </div>
    </div>
  </div>
</div>

`;

exports[`ui/shared/components/Home/PopularCities server renders apartments HTML 1`] = `

<h4>
  Apartments
</h4>
<div>
  <ul>
    <li>
      <a href="/georgia/atlanta-apartments">
        Atlanta, GA
      </a>
    </li>
    <li>
      <a href="/north-carolina/charlotte-apartments">
        Charlotte, NC
      </a>
    </li>
    <li>
      <a href="/illinois/chicago-apartments">
        Chicago, IL
      </a>
    </li>
    <li>
      <a href="/california/los-angeles-apartments">
        Los Angeles, CA
      </a>
    </li>
    <li>
      <a href="/new-york/new-york-apartments">
        New York, NY
      </a>
    </li>
    <li>
      <a href="/michigan/detroit-apartments">
        Detroit, MI
      </a>
    </li>
    <li>
      <a href="/california/san-diego-apartments">
        San Diego, CA
      </a>
    </li>
    <li>
      <a href="/texas/dallas-apartments">
        Dallas, TX
      </a>
    </li>
  </ul>
</div>

`;

exports[`ui/shared/components/Home/PopularCities server renders houses HTML 1`] = `

<h4>
  Houses
</h4>
<div>
  <ul>
    <li>
      <a href="/georgia/atlanta-houses">
        Atlanta, GA
      </a>
    </li>
    <li>
      <a href="/north-carolina/charlotte-houses">
        Charlotte, NC
      </a>
    </li>
    <li>
      <a href="/illinois/chicago-houses">
        Chicago, IL
      </a>
    </li>
    <li>
      <a href="/california/los-angeles-houses">
        Los Angeles, CA
      </a>
    </li>
    <li>
      <a href="/new-york/new-york-houses">
        New York, NY
      </a>
    </li>
    <li>
      <a href="/michigan/detroit-houses">
        Detroit, MI
      </a>
    </li>
    <li>
      <a href="/california/san-diego-houses">
        San Diego, CA
      </a>
    </li>
    <li>
      <a href="/texas/dallas-houses">
        Dallas, TX
      </a>
    </li>
  </ul>
</div>

`;
